gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/reactive/policy/adapter/policy/PolicyAdapter.java

```diff
@@ -170,23 +144,35 @@ private Completable policyStream(SyncExecutionContext ctx, ExecutionPhase phase)
             .andThen(
                 Completable.defer(
                     () -> {
-                        // Replace the chunks of the request or response if response is not completed.
                         if (ctx.isInterrupted()) {
-                            // The response could be marked as completed is the policy invoke the stream.failWith(...) method.
+                            // The context can be interrupted if the policy has invoked the stream.failWith(...) method.
                             return Completable.complete();
                         }
 
                         if (newBuffer == null || newBuffer.length() == 0) {
                             return Completable.complete();
                         }
 
-                        if (phase == ExecutionPhase.REQUEST) {
-                            return ((SyncRequest) ctx.request()).setBody(Maybe.just(newBuffer));
-                        } else {
-                            return ((SyncResponse) ctx.response()).setBody(Maybe.just(newBuffer));
-                        }
+                        // Replace the chunks of the request or response if not interrupted.
+                        return setBody(ctx, phase, newBuffer);
                     }
                 )
             );
     }
+
+    private Flowable<Buffer> getSource(SyncExecutionContext ctx, ExecutionPhase phase) {
+        if (phase == ExecutionPhase.REQUEST) {
+            return ctx.request().content();
+        } else {
```

Comment:
Are we sure phase could only be RESPONSE if it is not REQUEST ?

Reply:
Same as below